Cassandra won't start when deployed on GKE with 1.4.0-SNAPSHOT

## Bug Report
<!--
Thanks for filing an issue! Before hitting the button, please answer these questions.
Fill in as much of the template below as you can. 
-->

**Describe the bug**
Cassandra fails to start with the current `main` as the process cannot create directories for data nor commit logs when deploying on GKE:

```
ERROR [main] 2021-09-09 16:21:28,902 CassandraDaemon.java:909 - Exception encountered during startup
java.lang.ExceptionInInitializerError: null
    at org.apache.cassandra.service.CassandraDaemon.setup(CassandraDaemon.java:260)
    at org.apache.cassandra.service.CassandraDaemon.activate(CassandraDaemon.java:763)
    at org.apache.cassandra.service.CassandraDaemon.main(CassandraDaemon.java:887)
Caused by: java.lang.IllegalStateException: Failed to mkdirs /opt/cassandra/data/data; unable to start server
    at org.apache.cassandra.config.DatabaseDescriptor.createAllDirectories(DatabaseDescriptor.java:1422)
    at org.apache.cassandra.db.commitlog.CommitLog.<init>(CommitLog.java:94)
    at org.apache.cassandra.db.commitlog.CommitLog.construct(CommitLog.java:77)
    at org.apache.cassandra.db.commitlog.CommitLog.<clinit>(CommitLog.java:64)
    ... 3 common frames omitted
```

Looking at the directories and symlinks, it looks like `/var/lib/cassandra` isn't mounted so that the `cassandra`  user can write to it:

![Capture d’écran 2021-09-09 à 18 42 48](https://user-images.githubusercontent.com/5096002/132730197-29110aeb-08a5-44cd-b5ef-5c09f303e471.png)
![Capture d’écran 2021-09-09 à 18 43 44](https://user-images.githubusercontent.com/5096002/132730209-4a8a4135-effb-49ed-b3df-d12acee80e9e.png)

**To Reproduce**
Deploy K8ssandra 1.4.0-SNAPSHOT on GKE using the following Helm values file:

```
cassandra:
  version: "4.0.0"
  heap:
   size: 4g
  gc:
    g1:
      enabled: true
      setUpdatingPauseTimePercent: 5
      maxGcPauseMillis: 300
  resources:
    requests:
      memory: "8Gi"
      cpu: "2000m"

  datacenters:
  - name: dc1
    size: 3
  ingress:
    enabled: false

  cassandraLibDirVolume:
    storageClass: standard-rwo
    size: 300Gi
```

**Expected behavior**
Cassandra should start and data volumes should mount so that required directories can be created by the `cassandra` user.
In K8ssandra 1.3.1, the rights for the `/var/lib/cassandra` are set differently, allowing Cassandra to start properly:
![Capture d’écran 2021-09-09 à 19 03 32](https://user-images.githubusercontent.com/5096002/132730683-190727ee-67a5-401f-b4db-0a33427e0de7.png)



┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-875) by [Unito](https://www.unito.io)
┆Issue Number: K8SSAND-875
┆Priority: Medium

--- a/charts/cass-operator/values.yaml
+++ b/charts/cass-operator/values.yaml
@@ -58,15 +58,12 @@ podSecurityContext: {}
 securityContext:
   # -- Mark root filesystem as read only
   readOnlyRootFilesystem: false
-
   # -- Run cass-operator container as non-root user
-  # runAsNonRoot: true
-
+  runAsNonRoot: true
   # -- Group for the user running the cass-operator container / process
-  # runAsGroup: 65534
-
+  runAsGroup: 65534
   # -- User for running the cass-operator container / process
-  # runAsUser: 65534
+  runAsUser: 65534
 
 # -- Resources requests and limits for the cass-operator pod. We usually
 # recommend not to specify default resources and to leave this as a conscious

--- a/charts/k8ssandra-common/templates/_helpers.tpl
+++ b/charts/k8ssandra-common/templates/_helpers.tpl
@@ -29,7 +29,7 @@ SecurityContext helpers for pod and container scope.
 k8ssandra container securityContext defaults
 **/}}
 {{- define "k8ssandra-common.container.security-ctx-defaults" -}}
-readOnlyRootFilesystem: true
+readOnlyRootFilesystem: false
 {{- end }}
 
 {{/**

--- a/charts/k8ssandra/templates/cassandra/cassdc.yaml
+++ b/charts/k8ssandra/templates/cassandra/cassdc.yaml
@@ -166,7 +166,6 @@ spec:
 {{- end }}
   podTemplateSpec:
     spec:
-      securityContext: {{- ternary ($defaultPodSecurityCtx) (.Values.cassandra.podSecurityContext | toYaml) (empty .Values.cassandra.podSecurityContext) | nindent 10 }}
       initContainers:
       - name: base-config-init
         image: {{ include "k8ssandra.cassandraImage" . }}

--- a/tests/unit/template_cassdc_test.go
+++ b/tests/unit/template_cassdc_test.go
@@ -159,7 +159,9 @@ var _ = Describe("Verify CassandraDatacenter template", func() {
 			AssertContainerSecurityContextExistsAndMatches(cassdc, CassandraContainer, expectedCtx)
 
 			// Default pod security context for cassdc
-			Expect(cassdc.Spec.PodTemplateSpec.Spec.SecurityContext).ToNot(BeNil())
+			// TODO - revisit as this was potentially causing issues by defaulting to {}
+			// with respect to running in GKE - see k8ssandra #1094
+			Expect(cassdc.Spec.PodTemplateSpec.Spec.SecurityContext).To(BeNil())
 		})
 
 		It("is not rendered if disabled", func() {
@@ -514,7 +516,7 @@ var _ = Describe("Verify CassandraDatacenter template", func() {
 		It("enabling reaper and medusa", func() {
 			// Simple verification that both have properties correctly applied
 			options := &helm.Options{
-				SetValues: map[string]string{"medusa.enabled": "true"},
+				SetValues:      map[string]string{"medusa.enabled": "true"},
 				KubectlOptions: defaultKubeCtlOptions,
 			}
 

--- a/tests/unit/utils/cassdc/containers.go
+++ b/tests/unit/utils/cassdc/containers.go
@@ -60,7 +60,9 @@ func AssertContainerSecurityContextExists(cassdc *cassop.CassandraDatacenter, na
 		Expect(container).ToNot(BeNil())
 		Expect(container.SecurityContext).ToNot(BeNil())
 		Expect(container.SecurityContext.ReadOnlyRootFilesystem).ToNot(BeNil())
-		Expect(*container.SecurityContext.ReadOnlyRootFilesystem).To(BeTrue())
+		// TODO - once images are adjusted that run as apps in these containers,
+		// this will be expecting read-only to be true.
+		Expect(*container.SecurityContext.ReadOnlyRootFilesystem).To(BeFalse())
 	}
 }
 

